### ISSUE TITLE
Folder logic enhancement: folder per function group

### ISSUE THREAD

**Issue by FreHu** [abapGit/abapGit#4000]
Function groups tend to contain a lot of includes with meaningless names, which makes your repo very messy without folders. Older code is often structured in function groups.

**Comment by mbtools** [comment]
or we serialize function groups like classes... all code in one file.

**Comment by FreHu** [comment]
This would be huge, also the includes usually correspond to forms so there is some sense to the structure, it's just hard to navigate.

**Comment by flaiker** [comment]
Newer code is also often structured in function groups as some functionality was never ported over to classes (rfc, update task, change documents, table maintenance, dynpro support, exits...).
The includes are, "easily", identifiable by their 3 character long suffix naming convention... Seriously though, I don't think a folder structure would help at all. You can only navigate larger function groups with IDE support where navigation to the relevant include is done automatically by selecting the component / modularization unit.

**Comment by larshp** [comment]
folders correspond to packages, having function groups on top of this would be confusing IMHO, but it could be a user setting

code should be serialized as close to what the developer sees, one include is one file

as a workaround, the developer can always move that one function group to a new package, or refactor the code to classes. Function groups are good for RFC, which is external communication, this would be nice to group in one package, and have all the business logic in a different package implemented with classes

**Comment by FreHu** [comment]
It's not that easy to change package hierarchies, there's more bureaucracy involved and this actually complicates our downport just because I want to create some subpackages and relocate a bunch of objects there. 

Well I am still doing that, but the plan for the function groups is to keep them where they are and slowly turn them into classes in the relevant subpackages. This might take a long time tohugh which is why I thought to solve it on the abapgit side.

Basically a function group on the system already looks like a folder, I just want it to follow the same principle in the repo. Of course this would be configurable just like full/prefix logic.

Btw many packages use dashes instead of underscores in names so they can't use the prefix logic, this would maybe also be a nice thing to support.

**Comment by larshp** [comment]
another idea, use https://docs.abapgit.org/ref-exits.html#change_tadir, to simulate the desired package layout

might need another/different exit for pull

**Comment by pokrakam** [comment]
It would make things tidier especially when packages have been littered with table maintenance generator objects, but I see this as a nice to have. 

On the one hand serialising into one file would be the cleanest solution, conceptually function modules are the equivalent of methods in a class. But on the other hand the way we usually work with FMs is such that we usually don't know or care what function group it's in - I just want to call function X - so I might also want to find it by name in the Git repo.

So on balance I think separating a FG out into a package is the easiest and cleanest. Shouldn't take more than a couple of minutes, or am I missing something? @FreHu what bureaucracy problems are you referring to? In most organisations shifting stuff into a subpackage is no issue.

**Comment by FreHu** [comment]
I don't have authorization to create packages :(. In sap for some things you need an overseer with special rights or to take a course or something. I would assume that some of our customers are similarly agile in their processes :)

This has some fun consequences - I've seen my colleague create an object with the wrong name and he didn't have authorization to delete it afterwards. I also have two transport requests I created in the wrong place that I'm not authorized to delete.

I'm gonna give up on this idea as there doesn't seem that much appreciation for it and it complicates things just to make the repo slightly prettier. It's probably better to pursue other things.

**Comment by pokrakam** [comment]
Urgh, must admit I've never worked with such restrictions, my sympathies. They obviously want control to keep order, so appeal to their sensibilities and suggest generated classes just pollute own-developed packages and should be moved into a subpackage... you may even get someone to do the work for you :-)

**Comment by mbtools** [comment]
nothing matters except $$$. translate the bureaucracy into time waste times your hourly rate. then customers will at least listen.

**Comment by mbtools** [comment]
not happening
